htlcswitch/interceptable_switch.go

```diff
@@ -30,80 +30,279 @@ var (
+
+		select {
+		case interceptor := <-s.interceptorRegistration:
+			s.setInterceptor(interceptor)
+
+		case packets := <-s.intercepted:
+			var notIntercepted []*htlcPacket
+			for _, p := range packets {
+				if s.interceptor == nil ||
+					!s.interceptForward(p) {
+
+					log.Debugf("Skipping interception of %v:%v",
+						p.incomingChanID, p.incomingHTLCID)
+
+					notIntercepted = append(notIntercepted, p)
+				}
+			}
+			return s.htlcSwitch.ForwardPackets(
+				nil, notIntercepted...,
+			)
+
+		case res := <-s.resolutionChan:
+			res.errChan <- s.resolve(res.resolution)
+
+		case <-s.quit:
+			return nil
+		}
+	}
+}
+
+func (s *InterceptableSwitch) resolve(res *FwdResolution) error {
+	log.Debugf("Resolving intercepted forward %v", res.Key)
+
+	intercepted, ok := s.holdForwards[res.Key]
+	if !ok {
+		return fmt.Errorf("fwd %v not found", res.Key)
+	}
+	delete(s.holdForwards, res.Key)
+
+	switch res.Action {
+	case FwdActionResume:
+		return intercepted.resume()
+
+	case FwdActionSettle:
+		return intercepted.settle(res.Preimage)
+
+	case FwdActionFail:
+		if len(res.FailureMessage) > 0 {
+			return intercepted.fail(res.FailureMessage)
+		} else {
+			return intercepted.failWithCode(res.FailureCode)
+		}
+
+	default:
+		return fmt.Errorf("unrecognized action %v", res.Action)
+	}
+}
+
+type FwdAction int
+
+const (
+	FwdActionResume FwdAction = iota
+	FwdActionSettle
+	FwdActionFail
+)
+
+type FwdResolution struct {
+	Key channeldb.CircuitKey
+
+	Action         FwdAction
+	Preimage       lntypes.Preimage
+	FailureMessage []byte
+	FailureCode    lnwire.FailCode
+}
+
+type fwdResolution struct {
+	resolution *FwdResolution
+	errChan    chan error
+}
+
+func (s *InterceptableSwitch) Resolve(res *FwdResolution) error {
+	internalRes := &fwdResolution{
+		resolution: res,
+		errChan:    make(chan error, 1),
+	}
+
+	select {
+	case s.resolutionChan <- internalRes:
+
+	case <-s.quit:
+		return errors.New("switch shutting down")
+	}
+
+	select {
+	case err := <-internalRes.errChan:
+		return err
+
+	case <-s.quit:
+		return errors.New("switch shutting down")
+	}
 }
 
 // ForwardPackets attempts to forward the batch of htlcs through the
 // switch, any failed packets will be returned to the provided
 // ChannelLink. The link's quit signal should be provided to allow
 // cancellation of forwarding during link shutdown.
-func (s *InterceptableSwitch) ForwardPackets(linkQuit chan struct{},
+func (s *InterceptableSwitch) ForwardPackets(_ chan struct{},
 	packets ...*htlcPacket) error {
 
-	var interceptor ForwardInterceptor
-	s.Lock()
-	interceptor = s.fwdInterceptor
-	s.Unlock()
+	// Synchronize with the main event loop. This should be light in the
+	// case where there is no interceptor.
+	log.Debugf("Interceptable switch received %v packets", len(packets))
 
-	// Optimize for the case we don't have an interceptor.
-	if interceptor == nil {
-		return s.htlcSwitch.ForwardPackets(linkQuit, packets...)
-	}
+	select {
+	case s.intercepted <- packets:
```

Comment:
Need to think how the onchain interception flow (https://github.com/lightningnetwork/lnd/pull/6219) can connect to this.

Reply:
I rebased #6219 to demonstrate how it can be done.